lib/cache/PackFileCacheStrategy.js

```diff
@@ -112,170 +124,448 @@ class Pack {
+		/** @type {number[]} */
+		const smallUsedContents = [];
+		/** @type {number} */
+		let smallUsedContentSize = 0;
+		/** @type {number[]} */
+		const smallUnusedContents = [];
+		/** @type {number} */
+		let smallUnusedContentSize = 0;
+		for (let i = 0; i < this.content.length; i++) {
+			const content = this.content[i];
+			if (content === undefined) continue;
+			if (content.outdated) continue;
+			const size = content.getSize();
+			if (size < 0 || size > MIN_CONTENT_SIZE) continue;
+			if (content.used.size > 0) {
+				smallUsedContents.push(i);
+				smallUsedContentSize += size;
 			} else {
-				const packEntry = new PackEntry(data, identifier);
-				const lastSize = this.lastSizes.get(identifier);
-				if (lastSize > MAX_INLINE_SIZE) {
-					write(() => packEntry);
-				} else {
-					write(packEntry);
-				}
+				smallUnusedContents.push(i);
+				smallUnusedContentSize += size;
 			}
 		}
-		write(null);
+
+		let mergedContent;
+		// 2. Check if minimium number is reached
+		if (
+			smallUsedContents.length >= CONTENT_COUNT_TO_MERGE ||
+			smallUsedContentSize > MIN_CONTENT_SIZE
+		) {
+			mergedContent = smallUsedContents.map(i => this.content[i]);
+
+			// 3. Remove old content entries
+			for (const i of smallUsedContents) {
+				this.content[i] = undefined;
+			}
+		} else if (
+			smallUnusedContents.length >= CONTENT_COUNT_TO_MERGE ||
+			smallUnusedContentSize > MIN_CONTENT_SIZE
+		) {
+			mergedContent = smallUnusedContents.map(i => this.content[i]);
+
+			// 3. Remove old content entries
+			for (const i of smallUnusedContents) {
+				this.content[i] = undefined;
+			}
+		} else return;
+
+		// 4. Determine merged items
+		/** @type {Set<string>} */
+		const mergedItems = new Set();
+		/** @type {Set<string>} */
+		const mergedUsedItems = new Set();
+		for (const content of mergedContent) {
+			for (const identifier of content.items) {
+				mergedItems.add(identifier);
+			}
+			for (const identifer of content.used) {
+				mergedUsedItems.add(identifer);
+			}
+		}
+
+		// 5. GC and update location of merged items
+		const newLoc = this._findLocation();
+		this._gcAndUpdateLocation(mergedItems, mergedUsedItems, newLoc);
+
+		// 6. If not empty, store content somewhere
+		if (mergedItems.size > 0) {
+			this.content[newLoc] = new PackContent(
+				mergedItems,
+				mergedUsedItems,
+				memorize(async () => {
+					// unpack existing content in parallel
+					// after that values are accessible in .content
+					await Promise.all(mergedContent.map(c => c.unpack()));
+					/** @type {Map<string, any>} */
+					const map = new Map();
+					for (const content of mergedContent) {
+						for (const [identifier, value] of content.content) {
+							if (mergedItems.has(identifier)) {
+								map.set(identifier, value);
+							}
+						}
+					}
+					return map;
+				})
+			);
+			this.logger.log(
+				"Merged %d small files with %d cache items into pack %d",
+				mergedContent.length,
+				mergedItems.size,
```

Comment:
As I understand this info could be mistaken since there is check before merging(line 293):
```
							if (mergedItems.has(identifier)) {
								map.set(identifier, value);
							}
```